File: python/modcdp/launcher/BBBrowserLauncher.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/launcher/BBBrowserLauncher.ts
# - ./go/modcdp/launcher/BBBrowserLauncher.go
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from typing import Any
from websocket import create_connection

from ..launcher.BrowserLauncher import LauncherConfig, BrowserLauncher, LaunchedBrowser, _launcher_config


DEFAULT_BROWSERBASE_VIEWPORT = {"width": 1288, "height": 711}


class BBBrowserLauncher(BrowserLauncher):
    def __init__(self, config: LauncherConfig | dict | None = None) -> None:
        raw_config = config.model_dump() if isinstance(config, LauncherConfig) else dict(config or {})
        super().__init__({**raw_config, "launcher_mode": "bb"})

    def launch(self, config: LauncherConfig | dict | None = None) -> LaunchedBrowser:
        merged = self.config if config is None else _launcher_config({**self.config.model_dump(), **_launcher_config(config).model_dump(exclude_unset=True)})
        browserbase_api_key = _first_string(merged.launcher_bb_api_key, os.environ.get("BROWSERBASE_API_KEY"))
        if not browserbase_api_key:
            raise RuntimeError("launcher_mode=bb requires BROWSERBASE_API_KEY or launcher.launcher_bb_api_key.")

        base_url = merged.launcher_bb_base_url
        resume_session_id = _first_string(merged.launcher_bb_session_id)
        keep_alive = _first_bool(merged.launcher_bb_keep_alive) or False
        close_session_on_close = _first_bool(merged.launcher_bb_close_session_on_close)
        if close_session_on_close is None:
            close_session_on_close = not keep_alive

        created_session = False
        if resume_session_id:
            session = _browserbase_request(
                base_url=base_url,
                browserbase_api_key=browserbase_api_key,
                method="GET",
                pathname=f"/v1/sessions/{resume_session_id}",
            )
        else:
            session_create_params = _object_value(merged.launcher_bb_session_create_params)
            browser_settings = {
                **_object_value(session_create_params.get("browserSettings")),
                **_object_value(merged.launcher_bb_browser_settings),
            }
            user_metadata = {
                **_object_value(session_create_params.get("userMetadata")),
                **_object_value(merged.launcher_bb_user_metadata),
            }
            extension_id = _first_string(
                merged.launcher_bb_extension_id,
                session_create_params.get("extensionId"),
                _object_value(session_create_params.get("browserSettings")).get("extensionId"),
            )
            region = _first_string(merged.launcher_bb_region, session_create_params.get("region"))
            body: dict[str, Any] = {
                **session_create_params,
                **({"keepAlive": True} if keep_alive else {}),
                **({"region": region} if region else {}),
                **({"timeout": merged.launcher_bb_timeout} if isinstance(merged.launcher_bb_timeout, int) else {}),
                **({"extensionId": extension_id} if extension_id else {}),
                "browserSettings": {
                    **browser_settings,
                    **({"extensionId": extension_id} if extension_id else {}),
                    "viewport": browser_settings.get("viewport"),
                },
                "userMetadata": {
                    **user_metadata,
                    "modcdp": "true",
                },
            }
            session = _browserbase_request(
                base_url=base_url,
                browserbase_api_key=browserbase_api_key,
                method="POST",
                pathname="/v1/sessions",
                body=body,
            )
            created_session = True

        session_id = _first_string(session.get("id"))
        connect_url = _first_string(session.get("connectUrl"))
        if not session_id or not connect_url:
            raise RuntimeError("Browserbase session creation returned an unexpected shape.")

        closed = False

        def close() -> None:
            nonlocal closed
            if closed:
                return
            closed = True
            if not created_session or not close_session_on_close:
                return
            _close_browser_cdp(connect_url)
            try:
                _browserbase_request(
                    base_url=base_url,
                    browserbase_api_key=browserbase_api_key,
                    method="POST",
                    pathname=f"/v1/sessions/{session_id}",
                    body={"status": "REQUEST_RELEASE"},
                )
            except Exception:
                pass

        self.launched = LaunchedBrowser(
            # Browserbase connectUrl is already a WebSocket CDP endpoint.
            cdp_url=connect_url,
            browserbase_session_id=session_id,
            browserbase_session_url=f"https://www.browserbase.com/sessions/{session_id}",
            browserbase_debug_url=_first_string(session.get("debuggerUrl"), session.get("debuggerFullscreenUrl")),
            close=close,
        )
        return self.launched


def _browserbase_request(
    *,
    base_url: str,
    browserbase_api_key: str,
    method: str,
    pathname: str,
    body: dict[str, Any] | None = None,
) -> dict[str, Any]:
    request = urllib.request.Request(
        _browserbase_url(base_url, pathname),
        method=method,
        headers={
            "content-type": "application/json",
            "x-bb-api-key": browserbase_api_key,
        },
        data=json.dumps(body).encode() if body is not None else None,
    )
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            return json.loads(response.read())
    except urllib.error.HTTPError as error:
        text = error.read().decode(errors="replace")
        raise RuntimeError(f"Browserbase {method} {pathname} -> {error.code}{f': {text}' if text else ''}") from error


def _browserbase_url(base_url: str, pathname: str) -> str:
    return f"{base_url.rstrip('/')}/{pathname.lstrip('/')}"


def _close_browser_cdp(cdp_url: str | None) -> None:
    if not cdp_url or not cdp_url.startswith(("ws://", "wss://")):
        return
    try:
        ws = create_connection(cdp_url, timeout=2)
        try:
            ws.send(json.dumps({"id": 1, "method": "Browser.close", "params": {}}))
            ws.settimeout(0.5)
            try:
                ws.recv()
            except Exception:
                pass
        finally:
            ws.close()
    except Exception:
        pass


def _first_string(*values: Any) -> str | None:
    for value in values:
        if isinstance(value, str) and value.strip():
            return value.strip()
    return None


def _first_bool(*values: Any) -> bool | None:
    for value in values:
        if isinstance(value, bool):
            return value
    return None


def _object_value(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}

File: go/modcdp/launcher/BBBrowserLauncher.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/launcher/BBBrowserLauncher.ts
// - ./python/modcdp/launcher/BBBrowserLauncher.py
package launcher

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultBrowserbaseLauncherBaseURL = "https://api.browserbase.com"

var defaultBrowserbaseViewport = map[string]any{"width": 1288, "height": 711}

type BBBrowserLauncher struct {
	BrowserLauncher
}

type browserbaseSession struct {
	ID                    string `json:"id"`
	ConnectURL            string `json:"connectUrl"`
	DebuggerURL           string `json:"debuggerUrl"`
	DebuggerFullscreenURL string `json:"debuggerFullscreenUrl"`
	Status                string `json:"status"`
}

func NewBBBrowserLauncher(config LauncherConfig) *BBBrowserLauncher {
	config.LauncherMode = "bb"
	return &BBBrowserLauncher{BrowserLauncher: NewBrowserLauncher(config)}
}

func (l *BBBrowserLauncher) Launch(config LauncherConfig) (*LaunchedBrowser, error) {
	merged := mergeLaunchConfig(l.Config, config)
	browserbaseAPIKey := firstString(merged.LauncherBBAPIKey, os.Getenv("BROWSERBASE_API_KEY"))
	if browserbaseAPIKey == "" {
		return nil, fmt.Errorf("launcher_mode=bb requires BROWSERBASE_API_KEY or launcher.launcher_bb_api_key.")
	}

	baseURL := merged.LauncherBBBaseURL
	resumeSessionID := firstString(merged.LauncherBBSessionID)
	keepAlive := boolValue(merged.LauncherBBKeepAlive, false)
	closeSessionOnClose := boolValue(merged.LauncherBBCloseSessionOnClose, !keepAlive)

	createdSession := false
	var session browserbaseSession
	var err error
	if resumeSessionID != "" {
		err = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodGet, "/v1/sessions/"+resumeSessionID, nil, &session)
	} else {
		sessionCreateParams := objectValue(merged.LauncherBBSessionCreateParams)
		browserSettings := mergeMap(objectValue(sessionCreateParams["browserSettings"]), objectValue(merged.LauncherBBBrowserSettings))
		userMetadata := mergeMap(objectValue(sessionCreateParams["userMetadata"]), objectValue(merged.LauncherBBUserMetadata))
		extensionID := firstString(
			merged.LauncherBBExtensionID,
			stringValue(sessionCreateParams["extensionId"]),
			stringValue(objectValue(sessionCreateParams["browserSettings"])["extensionId"]),
		)
		body := mergeMap(map[string]any{}, sessionCreateParams)
		if keepAlive {
			body["keepAlive"] = true
		}
		if region := firstString(merged.LauncherBBRegion, stringValue(sessionCreateParams["region"])); region != "" {
			body["region"] = region
		}
		if merged.LauncherBBTimeout != 0 {
			body["timeout"] = merged.LauncherBBTimeout
		}
		if extensionID != "" {
			body["extensionId"] = extensionID
			browserSettings["extensionId"] = extensionID
		}
		if !hasViewportWidth(objectValue(browserSettings["viewport"])) {
			browserSettings["viewport"] = defaultBrowserbaseViewport
		}
		userMetadata["modcdp"] = "true"
		body["browserSettings"] = browserSettings
		body["userMetadata"] = userMetadata
		err = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodPost, "/v1/sessions", body, &session)
		createdSession = true
	}
	if err != nil {
		return nil, err
	}
	if session.ID == "" || session.ConnectURL == "" {
		return nil, fmt.Errorf("Browserbase session creation returned an unexpected shape")
	}

	closed := false
	close := func() {
		if closed {
			return
		}
		closed = true
		if !createdSession || !closeSessionOnClose {
			return
		}
		closeBrowserCDP(session.ConnectURL)
		body := map[string]any{"status": "REQUEST_RELEASE"}
		var ignored map[string]any
		_ = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodPost, "/v1/sessions/"+session.ID, body, &ignored)
	}

	launched := &LaunchedBrowser{
		// Browserbase ConnectURL is already a WebSocket CDP endpoint.
		CDPURL:                session.ConnectURL,
		BrowserbaseSessionID:  session.ID,
		BrowserbaseSessionURL: "https://www.browserbase.com/sessions/" + session.ID,
		BrowserbaseDebugURL:   firstString(session.DebuggerURL, session.DebuggerFullscreenURL),
		Close:                 close,
	}
	l.Launched = launched
	return launched, nil
}

func browserbaseRequest(baseURL string, browserbaseAPIKey string, method string, pathname string, body map[string]any, out any) error {
	var reader io.Reader
	if body != nil {
		bodyBytes, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(bodyBytes)
	}
	request, err := http.NewRequest(method, strings.TrimRight(baseURL, "/")+"/"+strings.TrimLeft(pathname, "/"), reader)
	if err != nil {
		return err
	}
	request.Header.Set("content-type", "application/json")
	request.Header.Set("x-bb-api-key", browserbaseAPIKey)
	client := &http.Client{Timeout: 30 * time.Second}
	response, err := client.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	responseBody, _ := io.ReadAll(response.Body)
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("Browserbase %s %s -> %d%s", method, pathname, response.StatusCode, errorText(responseBody))
	}
	if out == nil || len(responseBody) == 0 {
		return nil
	}
	return json.Unmarshal(responseBody, out)
}

func closeBrowserCDP(wsURL string) {
	if !strings.HasPrefix(wsURL, "ws://") && !strings.HasPrefix(wsURL, "wss://") {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	conn, _, _, err := ws.Dial(ctx, wsURL)
	if err != nil {
		return
	}
	defer conn.Close()
	_ = wsutil.WriteClientMessage(conn, ws.OpText, []byte(`{"id":1,"method":"Browser.close","params":{}}`))
	_ = conn.SetDeadline(time.Now().Add(500 * time.Millisecond))
	_, _, _ = wsutil.ReadServerData(conn)
}

func errorText(body []byte) string {
	if len(body) == 0 {
		return ""
	}
	return ": " + string(body)
}

func boolValue(value *bool, fallback bool) bool {
	if value == nil {
		return fallback
	}
	return *value
}

func firstMap(values ...map[string]any) map[string]any {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}

func objectValue(value any) map[string]any {
	if object, ok := value.(map[string]any); ok {
		return object
	}
	return map[string]any{}
}

func mergeMap(left map[string]any, right map[string]any) map[string]any {
	merged := map[string]any{}
	for key, value := range left {
		merged[key] = value
	}
	for key, value := range right {
		merged[key] = value
	}
	return merged
}

func stringValue(value any) string {
	if value, ok := value.(string); ok {
		return strings.TrimSpace(value)
	}
	return ""
}

func hasViewportWidth(viewport map[string]any) bool {
	if _, ok := viewport["width"]; ok {
		return true
	}
	return false
}
